### DISCUSSION TITLE
self.impl.set(...) - AttributeError: 'NoneType' object has no attribute 'set'

### DISCUSSION THREAD

**Discussion by manawasp** [sqlalchemy/sqlalchemy#7781] [Usage Questions] [answered]
Hello,

I got an issue where in my tests while migrating from `1.3` to `1.4` there is a behavior difference between by e2e and integrations tests,

Integration tests are running on an issue when calling my model (with custom attribute GUID), I noticed it only appear with custom type:

```
self = <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x7f0f0b6e84f0>, instance = <myp.models.campaigns.Campaign object at 0x7f0f02af1780>
value = UUID('d1b73a47-2622-40de-8d5e-2a23ae24f14d')

    def __set__(self, instance, value):
>       self.impl.set(
            instance_state(instance), instance_dict(instance), value, None
        )
E       AttributeError: 'NoneType' object has no attribute 'set'
```

When I put a breakpoint i can see from model the impl is not set:

```
(Pdb) Campaign.uid.__dict__
{'class_': <class 'myp.models.campaigns.Campaign'>, 'key': 'uid', 'impl': None, 'comparator': <sqlalchemy.orm.properties.ColumnProperty.Comparator object at 0x7ff51b8d16c0>, '_of_type': None, '_extra_criteria': (), '__doc__': None}
```

But when I put a breakpoint in my e2e tests and do the same:

```
(Pdb) Campaign.uid.__dict__
{'class_': <class 'myp.models.campaigns.Campaign'>, 'key': 'uid', 'impl': <sqlalchemy.orm.attributes.ScalarAttributeImpl object at 0x7f4a6535a7a0>, 'comparator': <sqlalchemy.orm.properties.ColumnProperty.Comparator object at 0x7f4a6a657c40>, '_of_type': None, '_extra_criteria': (), '__doc__': None, 'dispatch': <sqlalchemy.event.base.AttributeEventsDispatch object at 0x7f4a65353010>}
```

I can see the `impl` well init. I think I miss something in the init step maybe ?

Thanks for reading/helping :pray: 

EDIT:

Campaign model example:

```python
from sqlalchemy.orm import declarative_base
Base = declarative_base()

class Campaign(Base):
    __tablename__ = "Campaigns"

    # generic properties

    id = Column(
        "Id",
        Integer,
        primary_key=True,
        nullable=False,
    )
    uid = Column(
        "Uid",
        GUID,
        nullable=False,
        default=uuid4,
    )
 ```
 
 GUID is took from the [doc here](https://docs.sqlalchemy.org/en/14/core/custom_types.html#backend-agnostic-guid-type)

**Comment by zzzeek** [comment, marked as answer]
try calling [configure_mappers](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=configure_mappers#sqlalchemy.orm.configure_mappers) at the point at which all your mappings are set up.

**Reply by zzzeek** [reply]
also if your tests are doing any kind of clear_mappers(), that would require some extra attention to make sure setup/teardown isn't interfering with subsequent tests

**Reply by manawasp** [reply]
Thanks ! It's worked well after added the `configure_mappers` in the right fixture :pray:
